ros/api/common/add_group_filter.py

```diff
@@ -7,16 +7,33 @@
 
 
 def group_filtered_query(query):
+    all_host_groups = get_host_groups()
+    host_groups_without_none = [group for group in all_host_groups if group is not None]
     if get_flag_value(FLAG_INVENTORY_GROUPS):
-        if len(get_host_groups()) >= 1:
-            query = query.filter(System.groups[0]['id'].astext.in_(get_host_groups()) | (System.groups == '[]'))
+        if len(all_host_groups) >= 1:
+            if load_ungrouped_sys(all_host_groups):
+                query = query.filter(System.groups[0]['id'].astext.in_(host_groups_without_none)
+                                     | (System.groups == '[]'))
+            elif not load_ungrouped_sys(all_host_groups):
+                query = query.filter(System.groups[0]['id'].astext.in_(host_groups_without_none))
+        elif len(all_host_groups) == 1 and load_ungrouped_sys(all_host_groups):
+            query = query.filter((System.groups == '[]'))
```

Comment:
```suggestion
 def group_filtered_query(query):
    if get_flag_value(FLAG_INVENTORY_GROUPS):
        total_groups_from_request = get_host_groups()
        len_of_total_groups = len(total_groups_from_request)
        no_none_groups = [grp for grp in total_groups_from_request if grp is not None]
        len_of_no_none_groups = len(no_none_groups)
        is_none_present = (len_of_total_groups - len_of_no_none_groups) > 0 
        filters = []

        if len_of_total_groups:
            if len_of_no_none_groups:
                filters.append(System.groups[0]['id'].astext.in_(no_none_groups))
            if is_none_present:
                filters.append(System.groups == '[]')

            if len(filters) > 0:
                query = query.filter(or_(*filters))            
    return query
    
```
Remove unnecessary if..elseif block and methods.
You can make a use of existing length values and with above suggestion, it will be simpler to read.

Reply:
Far better! thank you!